lib/dialects/mysql/index.js

```
@@ -191,6 +198,60 @@ class Client_MySQL extends Client {
       options: {},
     });
   }
+
+  async checkVersion(connection) {
+    if (this._versionCheckPromise) {
+      return this._versionCheckPromise;
+    }
+
+    this._versionCheckPromise = (async () => {
+      try {
+        const query = promisify(connection.query).bind(connection);
+        const rows = await query('select version() as version');
+        const rawVersion =
+          rows?.[0]?.version || rows?.[0]?.VERSION || rows?.[0]?.['VERSION()'];
+
+        if (!rawVersion) {
+          throw new Error('Unable to retrieve MySQL/MariaDB version');
+        }
```

Failure to parse a version is our fault, not the user's; I think, in this case, it would be better to allow the version to be "unknown" and produce a warning instead.

Since this information is used to positively gate features, failing closed means that the library can still be used, except that the gated feature(s) will be unavailable.

If there's an oversight in our code, or if something changes on e.g. mysql/mariadb's end that makes the parsing no longer succeed, then the scope of the failure is reduced from "can't use knex at all" to "can't use a specific feature"

---

Test

---

Please bear with me, I'm going to try and organize my thoughts and reduce back-and-forth :) 

## Regex version parsing

Regex is so easy to mess up, and version strings can change their shape and content over time. You see distro-level builds adding their own strings and formatting onto things, backported security fixes, and other random garbage. They're usually well-structured, but if we _don't know_ the structure that they're expected to have, we're just guessing. I've dealt with production incidents that came down to "we deployed a new version of this thing, and now it produces data that no longer matches our expectation".

If we're just guessing, my opinion is that we should ensure that we don't cause a total outage if our guess is wrong. If we're not guessing, we need to support that by demonstrating (with documentation, source code, or empirical data) that we have good coverage. I'm not happy with "one test on one version string from one instance of maridb" for the empirical approach.

## Value vs risk

The value of this code, in this file, at this time, is that we're able to run one new query: "drop index if exists". The risk is a total outage. That seems pretty one-sided to me.

Example: User upgrades (or security-patches) database. Version string no longer matches regex. Application refuses to connect. User now has an incident to deal with, where their only actionable options are to roll back their database or monkey patch a library dependency.

## Consistency across dialects

I agree with you that we should treat things consistently, and would be bullish on a PR to do that!

The reason _this_ dialect in _this_ change is different is that it's being merged into a codebase with like 2 million downloads a week and it hasn't been out in the wild yet. So I'm erring on the side of caution and thinking through the lens of "migrating code safely in place".

## Unexpected behavior

I agree that, in general, "something unknown is wrong" can be interpreted to mean "we can't rely on anything and we should make the user figure look into it". The difference here is that we know what's wrong, and we know its impact. What's wrong is "we failed to parse the version string". We aslo know:

- The user has told us we're talking to mysql/mariadb
- The version data is used to _opt in_ to some behavior

There is no code that I know of that would "open up features that shouldn't be opened up", and I'd be hesitant to write any feature-detected code that does so. Acknowledging that I may be missing some real cases, I expect code that checks database versions to be of the form: `if ( i know that this works ) { do the thing }`.

If you have something in mind here, let me know? I may well have missed something

---

In terms of "now", we have very little:

* Postgres: potentially the wrong DDL commands are issued with json/jsonb used/not used when they shouldn't/should be
* Oracle: DDLs might fail because of the 30 character limit

But what about the future? For Postgres we really should modernise how we make auto increment records. This is in existing issues. This would be behind a version check. I assume as we work on issues, we will  potentially gate more things behind version control. It's not a "now" problem but...

---

I'm in favor of kibertoad's suggestion of just letting failing queries fail, unless at some point we identify a feature that will _behave wrongly_. I don't know of any cases like that with what the current code is doing though; they're all syntax things that will just produce a syntax error(?)

As for an escape hatch, that need is covered by the "explicitly specify a version" behavior. If the user specifies a version, then knex behaves as though it's talking to that database version. You might support something like `latest` as an explicit "just wide open do whatever you can". This would at least keep from tacking on "yet another option to the bag".

The simplest rendering of the ideas presented here is:

- Allow user to specify a database version. If they do so, behave in accordance with that version, and perform no introspection.
- Escape hatch: If they explicitly specify 'latest', behave as though the database is of the latest version we know/care about.
- If the user does not specify a database version, query the database and parse the version to figure it out
- If unparseable, behave as though the user had specified 'latest'